Terminal.Gui/ConsoleDrivers/WindowsDriver.cs

```diff
@@ -875,14 +877,153 @@ private void ChangeWin (Object s, SizeChangedEventArgs e)
+		//case ConsoleKey.NumPad9:
+		//	return keyInfoEx.NumLock ? Key.D9 : Key.PageUp;
+
+		case ConsoleKey.Oem1:
+		case ConsoleKey.Oem2:
+		case ConsoleKey.Oem3:
+		case ConsoleKey.Oem4:
+		case ConsoleKey.Oem5:
+		case ConsoleKey.Oem6:
+		case ConsoleKey.Oem7:
+		case ConsoleKey.Oem8:
+		case ConsoleKey.Oem102:
+		case ConsoleKey.OemPeriod:
+		case ConsoleKey.OemComma:
+		case ConsoleKey.OemPlus:
+		case ConsoleKey.OemMinus:
+			return (Key)((uint)keyInfo.KeyChar);
+		}
+
+		var key = keyInfo.Key;
+
+		if (key >= ConsoleKey.A && key <= ConsoleKey.Z) {
+			var delta = key - ConsoleKey.A;
+			if (keyInfo.Modifiers == ConsoleModifiers.Control) {
+				return (Key)(((uint)Key.CtrlMask) | ((uint)Key.A + delta));
+			}
+			if (keyInfo.Modifiers == ConsoleModifiers.Alt) {
+				return (Key)(((uint)Key.AltMask) | ((uint)Key.A + delta));
+			}
+			if (keyInfo.Modifiers == (ConsoleModifiers.Shift | ConsoleModifiers.Alt)) {
+				return ConsoleKeyMapping.MapKeyModifiers (keyInfo, (Key)((uint)Key.A + delta));
+			}
+			if ((keyInfo.Modifiers & (ConsoleModifiers.Alt | ConsoleModifiers.Control)) != 0) {
+				if (keyInfo.KeyChar == 0 || (keyInfo.KeyChar != 0 && keyInfo.KeyChar >= 1 && keyInfo.KeyChar <= 26)) {
+					return ConsoleKeyMapping.MapKeyModifiers (keyInfo, (Key)((uint)Key.A + delta));
+				}
+			}
+
+			if (((keyInfo.Modifiers == ConsoleModifiers.Shift) ^ (keyInfoEx.CapsLock))) {
+				return (Key)((uint)Key.A + delta) | Key.ShiftMask;
+			}
+
+			return (Key)((uint)keyInfo.KeyChar);
+
+		}
+
+		if (key >= ConsoleKey.D0 && key <= ConsoleKey.D9) {
+			var delta = key - ConsoleKey.D0;
+			if (keyInfo.Modifiers == ConsoleModifiers.Alt) {
+				return (Key)(((uint)Key.AltMask) | ((uint)Key.D0 + delta));
+			}
+			if (keyInfo.Modifiers == ConsoleModifiers.Control) {
+				return (Key)(((uint)Key.CtrlMask) | ((uint)Key.D0 + delta));
+			}
+			if (keyInfo.Modifiers == (ConsoleModifiers.Shift | ConsoleModifiers.Alt)) {
+				return ConsoleKeyMapping.MapKeyModifiers (keyInfo, (Key)((uint)Key.D0 + delta));
+			}
+			if ((keyInfo.Modifiers & (ConsoleModifiers.Alt | ConsoleModifiers.Control)) != 0) {
+				if (keyInfo.KeyChar == 0 || keyInfo.KeyChar == 30 || keyInfo.KeyChar == ((uint)Key.D0 + delta)) {
+					return ConsoleKeyMapping.MapKeyModifiers (keyInfo, (Key)((uint)Key.D0 + delta));
+				}
+			}
+			return (Key)((uint)keyInfo.KeyChar);
+		}
+
+		if (key >= ConsoleKey.F1 && key <= ConsoleKey.F12) {
+			var delta = key - ConsoleKey.F1;
+			if ((keyInfo.Modifiers & (ConsoleModifiers.Shift | ConsoleModifiers.Alt | ConsoleModifiers.Control)) != 0) {
+				return ConsoleKeyMapping.MapKeyModifiers (keyInfo, (Key)((uint)Key.F1 + delta));
+			}
+
+			return (Key)((uint)Key.F1 + delta);
+		}
+
+		if (key == (ConsoleKey)16) { // Shift
+			return Key.Null | Key.ShiftMask;
+		}
+
+		if (key == (ConsoleKey)17) { // Ctrl
+			return Key.Null | Key.CtrlMask;
+		}
+
+		if (key == (ConsoleKey)18) { // Alt
+			return Key.Null | Key.AltMask;
+		}
```

Comment:
Commenting this solves the https://github.com/gui-cs/Terminal.Gui/pull/2927#discussion_r1419272797 issue.

Reply:
@tig plestry this changed code and you'll see that more things work on `WindowsDriver`. Of course the unit tests pass because it's related with `FakeDriver`, which my need some adjust to reflect this change.

```cs
			var keyInfo = ToConsoleKeyInfoEx (inputEvent.KeyEvent);
			var map = MapKey (keyInfo);

			if (inputEvent.KeyEvent.bKeyDown) {
				_altDown = keyInfo.ConsoleKeyInfo.Modifiers == ConsoleModifiers.Alt;
				// Avoid sending repeat keydowns
				OnKeyDown (new KeyEventArgs (map));
				if (keyInfo.ConsoleKeyInfo.KeyChar != 0) {
					var keyPressedEventArgs = new KeyEventArgs (map);

					OnKeyPressed (keyPressedEventArgs);
				}
			} else {
				var keyPressedEventArgs = new KeyEventArgs (map);

				// PROTOTYPE: This logic enables `Alt` key presses (down, up, pressed).
				// However, if while the 'Alt' key is down, if another key is pressed and
				// released, there will be a keypressed event for that and the
				// keypressed event for just `Alt` will be suppressed. 
				// This allows MenuBar to have `Alt` as a keybinding
				if (map != Key.AltMask) {
					if (keyInfo.ConsoleKeyInfo.Modifiers.HasFlag (ConsoleModifiers.Alt)) {
						if (_altDown) {
							_altDown = false;
							OnKeyUp (new KeyEventArgs (map));
							//OnKeyPressed (new (Key.AltMask));
						}

						_altDown = false;
					}
					// KeyUp of an Alt-key press. 
					OnKeyUp (keyPressedEventArgs);
					//OnKeyPressed (keyPressedEventArgs);
				} else {
					OnKeyUp (keyPressedEventArgs);
					//if (_altDown) {
					//	_altDown = false;
					//	OnKeyPressed (keyPressedEventArgs);
					//}
				}
			}

```

Reply:
> @tig please use this changed code and you'll see that more things work on `WindowsDriver`. Of course the unit tests pass because it's related with `FakeDriver`, which my need some adjust to reflect this change.
> 
> ```cs
> 			var keyInfo = ToConsoleKeyInfoEx (inputEvent.KeyEvent);
> 			var map = MapKey (keyInfo);
> 
> 			if (inputEvent.KeyEvent.bKeyDown) {
> 				_altDown = keyInfo.ConsoleKeyInfo.Modifiers == ConsoleModifiers.Alt;
> 				// Avoid sending repeat keydowns
> 				OnKeyDown (new KeyEventArgs (map));
> 				if (keyInfo.ConsoleKeyInfo.KeyChar != 0) {
> 					var keyPressedEventArgs = new KeyEventArgs (map);
> 
> 					OnKeyPressed (keyPressedEventArgs);
> 				}
> 			} else {
> 				var keyPressedEventArgs = new KeyEventArgs (map);
> 
> 				// PROTOTYPE: This logic enables `Alt` key presses (down, up, pressed).
> 				// However, if while the 'Alt' key is down, if another key is pressed and
> 				// released, there will be a keypressed event for that and the
> 				// keypressed event for just `Alt` will be suppressed. 
> 				// This allows MenuBar to have `Alt` as a keybinding
> 				if (map != Key.AltMask) {
> 					if (keyInfo.ConsoleKeyInfo.Modifiers.HasFlag (ConsoleModifiers.Alt)) {
> 						if (_altDown) {
> 							_altDown = false;
> 							OnKeyUp (new KeyEventArgs (map));
> 							//OnKeyPressed (new (Key.AltMask));
> 						}
> 
> 						_altDown = false;
> 					}
> 					// KeyUp of an Alt-key press. 
> 					OnKeyUp (keyPressedEventArgs);
> 					//OnKeyPressed (keyPressedEventArgs);
> 				} else {
> 					OnKeyUp (keyPressedEventArgs);
> 					//if (_altDown) {
> 					//	_altDown = false;
> 					//	OnKeyPressed (keyPressedEventArgs);
> 					//}
> 				}
> 			}
> ```

Reply:
When I first read this comment I thought I agreed. But now looking at my code in the PR I'm not sure what you are referring to. Can you please clarify?

Reply:
I mean about the checking for Alt key that was already used before and perform the respective action.

Reply:
> When I first read this comment I thought I agreed. But now looking at my code in the PR I'm not sure what you are referring to. Can you please clarify?

This should be like this:

```cs
		//if (key == (ConsoleKey)16) { // Shift
		//	return Key.Null | Key.ShiftMask;
		//}

		//if (key == (ConsoleKey)17) { // Ctrl
		//	return Key.Null | Key.CtrlMask;
		//}

		//if (key == (ConsoleKey)18) { // Alt
		//	return Key.Null | Key.AltMask;
		//}

		return ConsoleKeyMapping.MapKeyModifiers (keyInfo, (Key)((uint)keyInfo.KeyChar));

```